budget/app.py

```diff
@@ -1,14 +1,176 @@
+    if response.ok:
+      results = json.loads(response.text)['results']
+      user_ids = [result['member']['ownerId'] for result in results if not result['isAdmin']]
+    else:
+      response.raise_for_status()
+    for user_id in user_ids:
+      if user_id in teams_by_user_id:
+        teams_by_user_id[user_id].append(team_id)
+      else:
+        teams_by_user_id[user_id] = [team_id]
+  return teams_by_user_id
+
+
+def check_user_duplicates(teams_by_user_id):
+  '''Verify that no users occur in multiple teams'''
+
+  duplicates = {user_id:team_memberships
+    for (user_id,team_memberships) in teams_by_user_id.items()
+    if len(team_memberships) > 1
+  }
+
+  if not duplicates:
+    return ''  # return an empty message if there are no duplicates
+  else:
+    return '\n'.join([
+      f'Synapse user id {user_id} occurs in teams {", ".join(team_memberships)}'
+      for (user_id, team_memberships) in duplicates.items()
+    ])
+
+
+def compare_budgets_and_users(account_id, users):
+  '''Finds users who lack a budget
+
+  This checks budget names, which will use the format
+  f'service-catalog_{synapse_id}', against the user list,
+  returning a list of user ids that need a budget to be made
+  '''
+  budgets_client = get_client('budgets')
+
+  # get budgets
+  budgets = budgets_client.describe_budgets(
+    AccountId=account_id
+    ).get('Budgets')
+
+  # get just the budget names
+  service_catalog_budget_names = [
+    budget['BudgetName'] for budget in budgets
+    if budget['BudgetName'].startswith('service-catalog_')
+  ]
+
+  # derive user ids from budget names
+  service_catalog_budgets_user_ids = set([
+    budget_name.split('_')[1] for budget_name in service_catalog_budget_names
+  ])
+
+  users = set(users)
+
+  # check for differences in ids found in budgets names and user ids from team
+  user_ids_without_budget = []
+  budgets_to_remove = []
+  if service_catalog_budgets_user_ids != users:
+    user_ids_without_budget = list(users - service_catalog_budgets_user_ids)
+    budgets_to_remove = list(service_catalog_budgets_user_ids - users)
+
+  return user_ids_without_budget, budgets_to_remove
+
+
+def create_budget(synapse_id, team):
+  '''Creates an AWS budget for a synapse user id'''
+  # TODO implement
+  log.info('Placeholder -- this will make a budget')
+
+
+def create_budgets(user_ids_without_budget, teams_by_user_id):
+  '''Creates a budget for each synapse id'''
+  new_budgets_created = []
+  for synapse_id in user_ids_without_budget:
+    team = teams_by_user_id[synapse_id][0]
+    create_budget(synapse_id, team)
+    new_budgets_created.append(synapse_id)
+
+  return f'Budgets created: {"none" if not new_budgets_created else ", ".join(new_budgets_created)}'
+
+
+def delete_budgets(synapse_ids):
+  # TODO implement
+  budgets_removed = []
+  return f'Budgets removed for synapse_ids: {"none" if not budgets_removed else ", ".join(budgets_removed)}'
+
 
 def lambda_handler(event, context):
+  '''Lambda event handler'''
```

Comment:
Is there a test for "lambda_handler"?

Reply:
There will be. As indicated in the PR description, this PR is one in a series. I have chosen to break the work up into multiple PRs to make reviewing a bit more digestible.